sys/arm64/vmm/vmm_hyp.c

```diff
@@ -0,0 +1,800 @@
+
+	WRITE_SPECIALREG(cnthctl_el2, cnthctl_el2);
+	WRITE_SPECIALREG(cntkctl_el1, cntkctl_el1);
+	WRITE_SPECIALREG(cntvoff_el2, cntvoff_el2);
+
+	return (ret);
+}
+
+static uint64_t
+vmm_hyp_read_reg(uint64_t reg)
+{
+	switch(reg) {
+	case HYP_REG_ICH_VTR:
+		return (READ_SPECIALREG(ich_vtr_el2));
+	case HYP_REG_CNTHCTL:
+		return (READ_SPECIALREG(cnthctl_el2));
+	}
+
+	return (0);
+}
+
+static bool
+vmm_is_vpipt_cache(void)
+{
+	/* TODO: Implement */
+	return (0);
+}
+
+static int
+vmm_clean_s2_tlbi(void)
+{
+	dsb(ishst);
+	__asm __volatile("tlbi alle1is");
+
+	/*
+	 * If we have a VPIPT icache it will use the VMID to tag cachelines.
+	 * As we are changing the allocated VMIDs we need to invalidate the
+	 * icache lines containing all old values.
+	 */
+	if (vmm_is_vpipt_cache())
+		__asm __volatile("ic ialluis");
+	dsb(ish);
+
+	return (0);
+}
+
+static int
+vm_s2_tlbi_range(uint64_t vttbr, vm_offset_t sva, vm_size_t eva,
+    bool final_only)
+{
+	uint64_t end, r, start;
+	uint64_t host_vttbr;
+
+#define	TLBI_VA_SHIFT			12
+#define	TLBI_VA_MASK			((1ul << 44) - 1)
+#define	TLBI_VA(addr)			(((addr) >> TLBI_VA_SHIFT) & TLBI_VA_MASK)
+#define	TLBI_VA_L3_INCR			(L3_SIZE >> TLBI_VA_SHIFT)
+
+	/* Switch to the guest vttbr */
+	/* TODO: Handle Cortex-A57/A72 erratum 131936 */
+	host_vttbr = READ_SPECIALREG(vttbr_el2);
+	WRITE_SPECIALREG(vttbr_el2, vttbr);
+	isb();
+
+	/*
+	 * The CPU can cache the stage 1 + 2 combination so we need to ensure
+	 * the stage 2 is invalidated first, then when this has completed we
+	 * invalidate the stage 1 TLB. As we don't know which stage 1 virtual
+	 * addresses point at the stage 2 IPA we need to invalidate the entire
+	 * stage 1 TLB.
+	 */
+
+	start = TLBI_VA(sva);
+	end = TLBI_VA(eva);
+	for (r = start; r < end; r += TLBI_VA_L3_INCR) {
+		/* Invalidate the stage 2 TLB entry */
+		if (final_only)
+			__asm __volatile("tlbi	ipas2le1is, %0" : : "r"(r));
+		else
+			__asm __volatile("tlbi	ipas2e1is, %0" : : "r"(r));
+	}
+	/* Ensure the entry has been invalidated */
+	dsb(ish);
+	/* Invalidate the stage 1 TLB. */
+	__asm __volatile("tlbi vmalle1is");
+	dsb(ish);
+	isb();
+
+	/* Switch back t othe host vttbr */
+	WRITE_SPECIALREG(vttbr_el2, host_vttbr);
+	isb();
+
+	return (0);
+}
+
+static int
+vm_s2_tlbi_all(uint64_t vttbr)
+{
+	uint64_t host_vttbr;
+
+	/* Switch to the guest vttbr */
+	/* TODO: Handle Cortex-A57/A72 erratum 131936 */
+	host_vttbr = READ_SPECIALREG(vttbr_el2);
+	WRITE_SPECIALREG(vttbr_el2, vttbr);
+	isb();
+
+	__asm __volatile("tlbi vmalls12e1is");
+	dsb(ish);
+	isb();
+
+	/* Switch back t othe host vttbr */
+	WRITE_SPECIALREG(vttbr_el2, host_vttbr);
+	isb();
+
+	return (0);
+}
+
+static int
+vmm_dc_civac(uintptr_t start, uint64_t len)
+{
+	size_t line_size, end;
+	uint64_t ctr;
+
+	ctr = READ_SPECIALREG(ctr_el0);
+	line_size = sizeof(int) << CTR_DLINE_SIZE(ctr);
+	end = start + len;
+	dsb(ishst);
+	/* Clean and Invalidate the D-cache */
+	for (; start < end; start += line_size)
```

Comment:
Are start and len guaranteed to be line_size-aligned?

Reply:
Existing callers guarantee it, but why would that be required? The ARM ARM claims that there are no alignment restrictions on the VA given to `dc civac`.

Reply:
Two things:

1. There's the misfeature that cache maintenance instructions in Morello require the bounds to cover the *whole* cache line, and will take a bounds fault if not. This is overly restrictive for everything other than `DC IVA*` and results in ugly code like exists in the cache_handle_range macro to amplify bounds to cover the whole cache line. If you can guarantee here that the requests are aligned then you don't need to worry about it, but otherwise you do.

2. Consider start = 1, len = 2 with a cache line size of 2 (for simplicity). Then the loop only runs once, namely for the line that includes address 1, because by the time you add 2 to start you're at 3, which isn't less than the end. You need to round end up for this kind of loop to work in that case (and at that point better to round start down to be extremely clear what range each loop iteration acts on).

Reply:
I see, ok. I'll add some code to simply round down to alignment. It's better to just handle this case rather than to try and signal an error since we can't really use panic() here and callers can ignore errors.